### PR TITLE
Fix language vs. locale

### LINKED ISSUE #15457: Timestamps are not displayed in the preferred language
### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments. 

### Steps to reproduce

Hi,

I first noticed this [issue](https://github.com/nextcloud/mail/issues/1232) with the email app and just today that the [contacts app](https://github.com/nextcloud/contacts/issues/1096) and the file app show the same behavior.

screenshot

The timestamps within Nextcloud are displayed in German even if my preferred language is English.

screenshot

I am using NC 16.0 and all apps up to date (mail 0.14.0, contacts 3.1.1)

### PR DESCRIPTION
**Pull request by nickvergessen**
* Steps: set language to english and locale to german
* Go to files app and upload a file
* Expected "a few seconds ago" and on hover "21.03.2019 …"
* Actually "vor einigen sekunden" and on hover "21.03.2019 …"
* Injecting getLanguage instead of getLocale will make it "a few seconds ago" and on hover "2019/21/03 …"

I checked, and indeed moment JS does not allow to differenciate between language and locale. My "idea" would be, to get the localeData of OC.getLocale(), then set moment.locale() to OC.getLanguage() and then overwrite weekstart, ispm, format, ... with the localeData of getLocale, or the other way around.

Opinions @ChristophWurst @rullzer @MorrisJobke 

```
_config: {…}
​​
abbr: "en"
​​
calendar: {…}
lastDay: "[Yesterday at] LT"
lastWeek: "[Last] dddd [at] LT"
nextDay: "[Tomorrow at] LT"
nextWeek: "dddd [at] LT"
sameDay: "[Today at] LT"
sameElse: "L"
<prototype>: Object { … }
​​
dayOfMonthOrdinalParse: /\d{1,2}(th|st|nd|rd)/
global: false
ignoreCase: false
lastIndex: 0
multiline: false
source: "\\d{1,2}(th|st|nd|rd)"
sticky: false
unicode: false
<prototype>: Object { … }
​​
invalidDate: "Invalid date"
​​
longDateFormat: {…}
L: "MM/DD/YYYY"
LL: "MMMM D, YYYY"
LLL: "MMMM D, YYYY h:mm A"
LLLL: "dddd, MMMM D, YYYY h:mm A"
LT: "h:mm A"
LTS: "h:mm:ss A"
<prototype>: Object { … }
​​
meridiemParse: /[ap]\.?m?\.?/i
global: false
ignoreCase: true
lastIndex: 0
multiline: false
source: "[ap]\\.?m?\\.?"
sticky: false
unicode: false
<prototype>: Object { … }
​​
months: (12) […]
0: "January"
1: "February"
2: "March"
3: "April"
4: "May"
5: "June"
6: "July"
7: "August"
8: "September"
9: "October"
10: "November"
11: "December"
length: 12
<prototype>: Array []
​​
monthsShort: (12) […]
0: "Jan"
1: "Feb"
2: "Mar"
3: "Apr"
4: "May"
5: "Jun"
6: "Jul"
7: "Aug"
8: "Sep"
9: "Oct"
10: "Nov"
11: "Dec"
length: 12
<prototype>: Array []
​​
ordinal: function ordinal()​​
relativeTime: Object { future: "in %s", past: "%s ago", s: "a few seconds", … }
​​
week: {…}
dow: 0
doy: 6
<prototype>: Object { … }
​​
weekdays: (7) […]
0: "Sunday"
1: "Monday"
2: "Tuesday"
3: "Wednesday"
4: "Thursday"
5: "Friday"
6: "Saturday"
length: 7
<prototype>: Array []
​​
weekdaysMin: (7) […]
0: "Su"
1: "Mo"
​​​2: "Tu"
​​​3: "We"
​​​4: "Th"
​​​5: "Fr"
​​​6: "Sa"
​​​length: 7
​​​<prototype>: Array []
​​weekdaysShort: (7) […]
​​​0: "Sun"
​​​1: "Mon"
​​​2: "Tue"
​​​3: "Wed"
​​​4: "Thu"
​​​5: "Fri"
​​​6: "Sat"
length: 7
```

fixes #15457

### CODE REVIEW

**Comment by ChristophWurst** [comment]
> I checked, and indeed moment JS does not allow to differenciate between language and locale. My "idea" would be, to get the localeData of OC.getLocale(), then set moment.locale() to OC.getLanguage() and then overwrite weekstart, ispm, format, ... with the localeData of getLocale, or the other way around.

While it sounds error prone I think this could actually work. I'm wondering if we're the first that have this problem or if there is a known workaround we could follow.

**Comment by nickvergessen** [comment]
I implemented hat solution and it seems to work quite well.

File: core/js/js.js [hunk 1]
```diff
@@ -895,9 +895,33 @@ function initCore() {
...
+		moment.locale([
+			OC.getLocale(),
+			OC.getLocale().split('_')[0],
+			OC.getLanguage(),
+			OC.getLanguage().split('_')[0],
+			'en',
+		]);
+		return moment.localeData()._abbr;
+	}
+	localeData.parentLocale = getBestParentLocale();
```
Comment (ChristophWurst) [inline core/js/js.js]: I would just inline this function call

**Review by ChristophWurst — APPROVED** [review]
Code looks good and change makes sense otherwise

**Comment by MorrisJobke** [comment]
>```
>   {
>     "message": "TypeError: undefined is not an object (evaluating 'OC.getLocale().split')\nat core/js/js.js:9:32995",
>     "str": "TypeError: undefined is not an object (evaluating 'OC.getLocale().split')\nat core/js/js.js:9:32995"
>   }
>```

🙈

**Comment by nickvergessen** [comment]
Passes locally. Maybe someone from @nextcloud/javascript can help?

**Comment by nickvergessen** [comment]
Rebased, lets see if time fixed it.

**Comment by MorrisJobke** [comment]
Conflicts again :/

**Comment by nickvergessen** [comment]
rebased

**Comment by nickvergessen** [comment]
Still failing :man_shrugging:

**Comment by MeCias** [comment]
This issue is still present in NC 17

**Comment by wiswedel** [comment]
@rullzer This looks kind of stale but keeps getting reported everywhere. Can this please be considered still for 18?

**Comment by MeCias** [comment]
This issue is still present in NC 18

**Comment by skjnldsv** [comment]
Status?

**Comment by nickvergessen** [comment]
See first post

**Comment by skjnldsv** [comment]
> See first post

I was it, what about it?

**Comment by nickvergessen** [comment]
Well i proposed an idea and now we need to say yes or no to this hack

**Comment by skjnldsv** [comment]
> While it sounds error prone I think this could actually work. I'm wondering if we're the first that have this problem or if there is a known workaround we could follow.

I'm in too!

**Comment by rullzer** [comment]
I'm going to close this due to lack of activity.
Feel free to reopen if anybody wants to continue.

**Comment by mrbenjoi** [comment]
This is still a problem in NC 22.2.0.

Is there a chance for progress on this issue?

**Pull request closed without merging**